keps/sig-node/1995-container-notifier/README.md

```diff
@@ -0,0 +1,899 @@
+    // Each notifier must have a unique name within the Container.
+    // Upon creation of a PodNotification, Container(s) in the specified
+    // Pod with a matching notifier name will be executed.
+    // +optional
+    Notifiers []ContainerNotifier
+    ......
+}
+
+type ContainerNotifier struct {
+    // Name of the ContainerNotifier. Name must be a valid label key. More info:
+    // https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set.
+    // Each ContainerNotifier within a Container must have a unique name.
+    // Names with the prefix `k8s.io/` are reserved for Kubernetes-defined
+    // "well-known" notifiers.
+    // Names with other prefixes such as `example.com/` or unprefixed are custom names.
+    // Immutable.
+    Name string
+
+    // The action to take when a matching PodNotification is observed.
+    // Exactly-once execution of the action is NOT guaranteed, so it's user's
+    // responsibility to make sure the handler is idempotent to get consistent
+    // behavior.
+    // For example, when a quiesce command has been ran on a database, the
+    // database will stay quiesced, and any consecutive execution of the
+    // same quiesce command should not have additional impact.
+    Handler *ContainerNotifierHandler
+
+    // Number of seconds after which the notifier will time out.
+    // Default to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
+}
+
+// ContainerNotifierHandler specifies an action to take when a ContainerNotifier
+// has been triggered.
+type ContainerNotifierHandler struct {
+    // Exec specifies the action to take.
+    // +optional
+    Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
+}
+```
+
+##### PodNotification
+
+The PodNotification object is a request for execution a ContainerNotifier in a `Pod`. It will be an in-tree API object in the core API group so that Kubelet can watch and trigger the actions.
+
+All containers in the specified `Pod` which have the requested "ContainerNotifier" defined in their spec will have their corresponding "ContainerNotifierHandler" executed.
+
+Note that there is no guarantee of execution ordering between containers.
+
+```
+type PodNotification struct {
+    metav1.TypeMeta
+    // +optional
+    metav1.ObjectMeta
+    // Spec defines the behavior of a notification.
+    // +optional
+    Spec PodNotificationSpec
+    // Status defines the current state of a notification.
+    // +optional
+    Status PodNotificationStatus
+}
+
+type PodNotificationSpec struct {
+    // The name of the Pod to which notification will be addressed.
+    // Note that a PodNotificaion only notifies a Pod in the same namespace.
+    // The PodNotification fails immediately if the specified Pod does not exist.
+    // Required.
+    // Immutable.
+    PodName string
+
+
+    // The name of the ContainerNotifier to trigger.
```

Comment:
PodNotificationSpec only has PodName and ContainerNotifier name (which is only container-level unique). So if a pod has two containers, each of which has a ContainerNotifier with the same name, how does PodNotificationSpec distinguish between them?

Reply:
If two containers have the same ContainerNotifier, then the same command will be run in both containers; If the intention is to run different commands in two different containers, each should use a different ContainerNotifier name.

Reply:
Got it, as long as this is intended we should be good. Actually you already mentioned that "All containers in the specified `Pod` which have the requested "ContainerNotifier" defined in their spec will have their corresponding "ContainerNotifierHandler" executed".

Reply:
It notifies all containers with that named notifier